k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,128 @@
+package io.buoyant.k8s
+
+import com.twitter.finagle.http.{Request, Response}
+import com.twitter.finagle.{Path, Service}
+import com.twitter.util._
+import java.util.concurrent.atomic.AtomicReference
+
+case class IngressSpec(
+  name: Option[String],
+  namespace: Option[String],
+  fallbackBackend: Option[IngressPath] = None,
+  rules: Seq[IngressPath]
+)
+
+case class IngressPath(
+  host: Option[String] = None,
+  uri: Option[String] = None,
+  namespace: String,
+  svc: String,
+  port: String
+)
+
+object IngressCache {
+  type IngressState = Activity.State[Seq[IngressSpec]]
+
+  private[k8s] def isPrefix(pfx: String, uri: String): Boolean = {
+    val pfxSegments = pfx.split("/")
+    val uriSegments = uri.split("/")
+    pfxSegments.corresponds(uriSegments.take(pfxSegments.size)) { _ == _ }
+  }
+
+  private[k8s] def getMatchingPath(hostHeader: Option[String], requestPath: String, ns: Option[String], ingresses: Seq[IngressSpec]): Option[IngressPath] = {
+    ingresses.flatMap { ingressResource =>
+      val matchingPath = ingressResource.rules.find { rule =>
```

Comment:
TIOLI: consider moving the path matching logic into a method on the IngressPath class:
```
def matches(hostHeader: Option[String], requestPath: String): Boolean = ???
```

and then you can define a `getMatchingPath` method on the `IngressSpec` class and then finally this method can simply become something like

```
ingresses
  .toIterator // to make this lazy so we can stop after we find a match
  .flatMap(_.getMatchingPath(...)) // see if there's a match
  .take(1) // take the first match
  .toSeq.headOption // convert back to option
```